letter-shop.md

```diff
@@ -0,0 +1,255 @@
+```
+
+Now wait for about 30 seconds, until you see something like:
+```txt
+Got the fulfillment, you paid for your letter! Go get it at http://localhost:8000/RWtoGF_sOVoIH3Casd-nmApQ-Thzl03lH-cInXume_g
+```
+
+As the instructions say, visit that URL to get your letter! :)
+
+## Step 3: Paying proxy
+
+It's of course very cumbersome to cut and paste the condition from your browser to your command-line terminal each time you need to pay for something online, and then to cut and past back the fulfillment from your terminal to your browser once you paid. Therefore, the following paying proxy is useful, which parses the shop's payment instructions, executes them, retreives the paid content, and serves it on port 8001. Save it as `proxy.js`:
+```js
+const IlpPacket = require('ilp-packet')
+const Plugin = require('ilp-plugin-xrp-escrow')
+const http = require('http')
+const fetch = require('node-fetch')
+const uuid = require('uuid/v4')
+function base64url (buf) { return buf.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '') }
+
+const plugin = new Plugin({
+  secret: 'sndb5JDdyWiHZia9zv44zSr2itRy1',
+  account: 'rGtqDAJNTDMLaNNfq1RVYgPT8onFMj19Aj',
+  server: 'wss://s.altnet.rippletest.net:51233',
+  prefix: 'test.crypto.xrp.'
+})
+
+const pendingRes = {}
+
+function sendTransfer (obj) {
+  obj.id = uuid()
+  obj.from = plugin.getAccount()
+  // to
+  obj.ledger = plugin.getInfo().prefix
+  // amount
+  obj.ilp = base64url(IlpPacket.serializeIlpPayment({ amount: obj.amount, account: obj.to }))
+  // executionCondition
+  obj.expiresAt = new Date(new Date().getTime() + 1000000).toISOString()
+  return plugin.sendTransfer(obj).then(function () {
+    return obj.id
+  })
+}
+
+plugin.connect().then(function () {
+  plugin.on('outgoing_fulfill', function (transfer, fulfillment) {
+    console.log('outgoing fulfill', transfer, fulfillment, 'http://localhost:8000/' + fulfillment)
+    fetch('http://localhost:8000/' + fulfillment).then(function (inRes) {
+      return inRes.text()
+    }).then(function (body) {
+      pendingRes[transfer.id].end(body)
+    })
+  })
+
+  http.createServer(function (req, outRes) {
+    fetch('http://localhost:8000' + req.url).then(function (inRes) {
+      return inRes.text()
+    }).then(function (body) {
+      const parts = body.split(' ')
+      // Please send an Interledger payment to test.crypto.xrp.rrhnXcox5bEmZfJCHzPxajUtwdt772zrCW with amount 10 drops and condition Z8_mOwUpkAvI-DTGMKQ_kfCIh7QgcYGu0uuFgloGVr4
+      // 0      1    2  3           4       5  6                                                  7    8      9  10    11  12        13
+      if (parts[0] === 'Please') {
+        sendTransfer({
+          to: parts[6],
+          amount: parts[9],
+          executionCondition: parts[13]
+        }).then(function (transferId) {
+          console.log('transfer sent', transferId)
+          pendingRes[transferId] = outRes
```

Comment:
I think it would be cleaner to add an event listener on the plugin here instead of storing the responses in that global object

Reply:
one listener per transferId? and then remove it once called, using `arguments.callee`, I guess?

Reply:
ah, using a named function. ok, that works! :)